src/cp/cp_control_point.c

```diff
@@ -0,0 +1,873 @@
+        new_trigger_count = ++(cp_registry->trigger_count);
+        __wt_verbose_debug1(session, WT_VERB_CONTROL_POINT,
+          "Triggered: id=%" PRId32 ", crossing_count=%" PRIu64 ", trigger_count=%" PRIu64, id,
+          (uint64_t)new_crossing_count, (uint64_t)new_trigger_count);
+    } else {
+        __wt_verbose_debug3(session, WT_VERB_CONTROL_POINT,
+          "Not Triggered: id=%" PRId32 ", crossing_count=%" PRIu64 ", trigger_count=%" PRIu64, id,
+          (uint64_t)new_crossing_count, (uint64_t)cp_registry->trigger_count);
+        /* Not triggered. */
+        data = NULL;
+    }
+    return (data);
+}
+
+/*
+ * Per connection control point initialization.
+ */
+/* From examples/ex_control_points.c */
+/*
+ * Per connection control point "Main Start Printing".
+ */
+/* Per connection control point data type. */
+struct __wt_conn_control_point_data_main_start_printing {
+    WT_CONTROL_POINT iface;
+    WT_CONTROL_POINT_ACTION_WAIT_FOR_TRIGGER action_data;
+};
+
+/* Per connection control point init function. */
+/*
+ * __wt_conn_control_point_init_main_start_printing --
+ *     The per connection control point initialization function for control point "Main Start
+ *     Printing".
+ *
+ * @param session The session. @param cp_registry The per connection control point's control point
+ *     registry. @param cfg Configuration strings.
+ */
+WT_CONTROL_POINT *
+__wt_conn_control_point_init_main_start_printing(WT_SESSION_IMPL *session, const char **cfg)
+{
+    struct __wt_conn_control_point_data_main_start_printing *init_data;
+    WT_DECL_RET;
+
+    ret = __wt_calloc_one(session, &init_data);
+    if (WT_UNLIKELY(ret != 0))
+        return (NULL);
+
+    /* Initialize configuration parameters. */
+    WT_ERR(__wt_control_point_config_action_wait_for_trigger(
+      session, (WT_CONTROL_POINT *)init_data, cfg));
+    /* The predicate is "Always" therefore no predicate configuration parameters to initialize. */
+
+    /* Extra initialization required for action "Wait for trigger". */
+    __wt_control_point_action_init_wait_for_trigger(
+      session, "Main Start Printing", (WT_CONTROL_POINT *)init_data);
+
+err:
+    if (ret != 0)
+        __wt_free(session, init_data);
+
+    return ((WT_CONTROL_POINT *)init_data);
+}
+
+/*
+ * Per connection control point "Thread 0".
+ */
+/* Per connection control point data type. */
+struct __wt_conn_control_point_data_thread_0 {
+    WT_CONTROL_POINT iface;
+    WT_CONTROL_POINT_ACTION_WAIT_FOR_TRIGGER action_data;
+};
+
+/* Per connection control point init function. */
+/*
+ * __wt_conn_control_point_init_thread_0 --
+ *     The per connection control point initialization function for control point "Thread 0".
+ *
+ * @param session The session. @param cp_registry The per connection control point's control point
+ *     registry. @param cfg Configuration strings.
+ */
+WT_CONTROL_POINT *
+__wt_conn_control_point_init_thread_0(WT_SESSION_IMPL *session, const char **cfg)
```

Comment:
Is there a way we can generate these __wt_conn_control_point_init_thread_* functions (eg from a macro) rather that requiring repeated similar code?

Reply:
They can be generated by the not yet written script in wt-13541-test-cp-with-wt-13212.
This is supposed to be an example of how to use control points. We should not do anything unusual.